### PR TITLE
Tappx Bid Adapter: new bidder adapter added

### PR DESCRIPTION
**Pull request by prebidtappx**
## Type of change
- [x] New bidder adapter

## Description of change
Add new Tappx adapter.

- Test parameters for validating bids
```
{
  bidder: 'tappx',
  params: {
    host: "testing.ssp.tappx.com/rtb/v2/",
    tappxkey: "pub-1234-android-1234",
    endpoint: "ZZ1234PBJS",
    bidfloor: 0.005
  }
}
```

- Contact email of the adapter’s maintainer: prebid@tappx.com
- [x] official adapter submission
- Link to PR on the docs repo: https://github.com/prebid/prebid.github.io/pull/2661

### CODE REVIEW

**Comment by prebidtappx** [comment]
Hello,
we have fixed the issues raised in commit 32ea819 on c1f42ae

**Comment by prebidtappx** [comment]
Hello @aleksatr 

is there any news regarding the review?
We've seen **feature** and **minor** tags added, and we're not sure if it's pending to review or waiting to merge to master.

Thanks in advance

**Comment by aleksatr** [comment]
Hi @prebidtappx , please make sure your unit test coverage is above 80%, it is required for every new PBJS adapter to be above this threshold. From what I see your coverage is at 78%, to generate and view the code coverage reports:

```bash
gulp test-coverage
gulp view-coverage
```

**Comment by prebidtappx** [comment]
Hello @aleksatr 

we have updated our tests. Now it covers 85%.
We await your comments.

Thank you

File: modules/tappxBidAdapter.js [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+   * @return ServerRequest Info describing the request to the server.
+   */
+  buildRequests: function(validBidRequests, bidderRequest) {
+    // Host definition
+    HOST = getParameters(validBidRequests, 'params.host');
+    hostDomain = HOST.split('/', 1)[0];
+
+    const ENDPOINT = getParameters(validBidRequests, 'params.endpoint');
+    const TAPPXKEY = getParameters(validBidRequests, 'params.tappxkey');
+    const BIDFLOOR = getParameters(validBidRequests, 'params.bidfloor');
```
Comment (aleksatr) [inline modules/tappxBidAdapter.js:44]: I'd like to bring your attention to Price Floors Module in PBJS, https://docs.prebid.org/dev-docs/modules/floors.html, this module provides a systematic way for publishers to configure and adapters to read price floor data. Did you consider reading price floor data from this module (when available)?

File: modules/tappxBidAdapter.js [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+
+    // > App/Site object
+    if (getParameters(validBidRequests, 'params.app')) {
+      let app = {};
+      app.name = getParameters(validBidRequests, 'params.app.name');
+      app.bundle = getParameters(validBidRequests, 'params.app.bundle');
+      app.domain = getParameters(validBidRequests, 'params.app.domain');
+      publisher.name = getParameters(validBidRequests, 'params.app.publisher.name');
+      publisher.domain = getParameters(validBidRequests, 'params.app.publisher.domain');
+      tagid = `${app.name}_typeAdBanVid_${getOs()}`;
```
Comment (aleksatr) [inline modules/tappxBidAdapter.js:61]: isn't this redundant, given that you can reconstruct this on the backend using app object?

File: modules/tappxBidAdapter.js [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+      payload.app = app;
+      api[0] = getParameters(validBidRequests, 'params.api') ? getParameters(validBidRequests, 'params.api') : [3, 5];
+    } else {
+      let site = {};
+      site.name = window.location.hostname;
```
Comment (aleksatr) [inline modules/tappxBidAdapter.js:66]: are you considering the case when PBJS is running inside an iframe? Did you check what is provided by `refererInfo` inside `bidRequest` and if you can use that instead?

File: modules/tappxBidAdapter.js [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+      banner.h = h;
+      if ((bannerMediaType.sizes[0].indexOf(480) >= 0) && (bannerMediaType.sizes[0].indexOf(320) >= 0)) {
+        banner.pos = 7
+      } else {
+        banner.pos = 4
+      }
+
+      banner.api = api;
+
+      let format = {};
```
Comment (aleksatr) [inline modules/tappxBidAdapter.js:95]: do you really want this to be a hashmap or you meant array?

File: modules/tappxBidAdapter.js [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+
+    // COPPA compliance
+    if (config.getConfig('coppa') === true) {
+      regs.coppa = config.getConfig('coppa') === true ? 1 : 0;
+    }
+    // < GDPR
+
+    // > Payload
+    payload.id = (getParameters(validBidRequests, 'auctionId'));
+    payload.test = 1;
```
Comment (aleksatr) [inline modules/tappxBidAdapter.js:160]: do you want this hardcoded?

File: modules/tappxBidAdapter.js [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+    payload.at = 1;
+    payload.tmax = 500;
```
Comment (aleksatr) [inline modules/tappxBidAdapter.js:162]: why is this hardcoded, maybe you wanted to read it from `bidderRequest.timeout` ?

**Comment by aleksatr** [comment]
@prebidtappx It doesn't seem to me that this adapter supports multiple bid requests in a single call to `buildRequests()`, for example, if I have two ad units, both including tappx bidder, I don't think that this adapter differentiates between the two and creates a proper request to backend, for example:

```
var adUnits = [{
    code: 'div-gpt-ad-1460505748561-0',
    mediaTypes: {
        banner: {
            sizes: [[300, 250], [300, 600]],
        }
    },
    // Replace this object to test a new Adapter!
    bids: [
        {
            bidder: "tappx",
            params: {
                host: "testing.ssp.tappx.com/rtb/v2/",
                tappxkey: "pub-1234-android-1234",
                endpoint: "ZZ1234PBJS",
                bidfloor: 0.005
            }
        }
    ]

},
{
    code: 'div-gpt-ad-1460505748561-1',
    mediaTypes: {
        banner: {
            sizes: [[300, 250], [300, 600]],
        }
    },
    // Replace this object to test a new Adapter!
    bids: [
        {
            bidder: "tappx",
            params: {
                host: "testing.ssp.tappx.com/rtb/v2/",
                tappxkey: "pub-1234-android-1235",
                endpoint: "ZZ1234PBJS",
                bidfloor: 0.005
            }
        }
    ]

}
];
```

Would produce payload:

```
{"site":{"name":"127.0.0.1","bundle":"127.0.0.1","domain":"127.0.0.1"},"id":"59dcf572-a8f8-4dd3-94ef-ec3b98d41669","test":1,"at":1,"tmax":500,"bidder":"tappx","imp":[{"banner":{"w":300,"h":250,"pos":4,"api":{},"format":{"0":{"w":300,"h":250}}},"id":1,"tagid":"127.0.0.1_typeAdBanVid_unknown","secure":1,"bidfloor":0.005}],"device":{"os":"unknown","ip":"peer","ua":"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/88.0.4324.96 Safari/537.36","h":1440,"w":2560,"dnt":0,"language":"en","make":"Google Inc."},"params":{"host":"tappx.com","tappxkey":"pub-1234-android-1234","endpoint":"ZZ1234PBJS","bidfloor":0.005},"regs":{"gdpr":0,"ext":{}}}
```

and I don't see mentions of second ad unit at all (for example searching for params)

File: modules/tappxBidAdapter.js (hunk 1 above)
Comment (prebidtappx) [reply modules/tappxBidAdapter.js:44]: Initially, we prefer to design the adapter as simple as possible, without modules and similar to our OpenRTB requests. We are designing the video feature and maybe we will consider adding the floor module with that function.

File: modules/tappxBidAdapter.js (hunk 2 above)
Comment (prebidtappx) [reply modules/tappxBidAdapter.js:62]: No, we have publishers that the app domain doesn’t match to the app domain. I.e.: **myapp.com** is the `app.domain`, and **company.com** corresponds to `publisher.domain`

File: modules/tappxBidAdapter.js (hunk 3 above)
Comment (prebidtappx) [reply modules/tappxBidAdapter.js]: Following your advice, we have updated the adapter to fix this mistake in the commit [31d23ee](https://github.com/prebidtappx/Prebid.js/commit/31d23ee4a7e5f2e5973a1d59c264d87aeb4b9887)

File: modules/tappxBidAdapter.js (hunk 4 above)
Comment (prebidtappx) [reply modules/tappxBidAdapter.js:98]: Yes, we need a hashmap because we have a key value

File: modules/tappxBidAdapter.js (hunk 5 above)
Comment (prebidtappx) [reply modules/tappxBidAdapter.js:163]: We have updated the adapter to fix this mistake in the commit: [18cad19](https://github.com/prebidtappx/Prebid.js/commit/18cad19ce681c5487197fbd987459f4282c4dfd0)

File: modules/tappxBidAdapter.js (hunk 6 above)
Comment (prebidtappx) [reply modules/tappxBidAdapter.js]: We have updated the adapter to get the value from biderRequest.tiemout or to set it to 600 if it doesn’t come there. Commit: [94e2357](https://github.com/prebidtappx/Prebid.js/commit/94e235771d29a09862bd52a776b328e350bc35db)

**Comment by prebidtappx** [comment]
@aleksatr thank you for your comments.
About multiple bid requests, you’re right, we don’t support it. For us, one request have to be one bid, is for that because we didn’t implement it on our adapter. It’s a problem?
Thank you

**Comment by aleksatr** [comment]
> About multiple bid requests, you’re right, we don’t support it. For us, one request have to be one bid, is for that because we didn’t implement it on our adapter. It’s a problem?

buildRequests() returns an array of requests, so you can create one request per bid and return an array, you can look for an example in other bid adapters (rubiconBidAdapter, ixBidAdapter, ... for example). In the current state you have, you will basically have a mixed bag of params, where the first bid has precedence.

**Comment by prebidtappx** [comment]
Hi @aleksatr , we've updated out adapter adding support for multiple bids in one request. We've followed the examples taht you proposed and testing with an adunit like the one you tried and we do the multiple bids right.
The test coverage is still above 80%.
We've updated out .md file with all sizes we support.

But the CircleCI tests fail here, but it's not related with our adapter isn't it? can you help us with that?
Thank you

```
Edge 17.17134.0 (Windows 10.0.0): Executed 7449 of 7454 (skipped 5) SUCCESS (29.451 secs / 17.286 secs)
Chrome 79.0.3945 (Windows 10.0.0): Executed 7449 of 7454 (2 FAILED) (skipped 5) (44.231 secs / 23.034 secs)
Edge 18.18363.0 (Windows 10.0.0): Executed 7449 of 7454 (skipped 5) SUCCESS (35.084 secs / 20.225 secs)
IE 11.0.0 (Windows 10.0.0): Executed 7449 of 7454 (1 FAILED) (skipped 5) (29.697 secs / 20.683 secs)
Chrome 80.0.3987 (Windows 10.0.0): Executed 7449 of 7454 (skipped 5) SUCCESS (20.392 secs / 10.379 secs)
Firefox 73.0.0 (Windows 10.0.0): Executed 7449 of 7454 (skipped 5) SUCCESS (28.061 secs / 15.079 secs)
Firefox 72.0.0 (Windows 10.0.0): Executed 7449 of 7454 (skipped 5) SUCCESS (25.908 secs / 14.155 secs)
Safari 13.1.2 (Mac OS X 10.15.6): Executed 7449 of 7454 (skipped 5) SUCCESS (9.646 secs / 6.86 secs)
Safari 12.1.2 (Mac OS X 10.14.6): Executed 7449 of 7454 (skipped 5) SUCCESS (10.281 secs / 7.08 secs)
TOTAL: 3 FAILED, 67038 SUCCESS


1) Logger: best case + win tracker
     pubmatic analytics adapter when handling events
     Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.

2) bidCpmAdjustment: USD: Logger: best case + win tracker
     pubmatic analytics adapter when handling events
     Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.

3) should build utm data from local storage
     Prebid Manager Analytics Adapter build utm tag data
     Timeout of 2000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.
```

File: modules/tappxBidAdapter.js (hunk 5 above)
Comment (aleksatr) [reply modules/tappxBidAdapter.js]: I still see test hardcoded to `1`?

Comment (prebidtappx) [reply modules/tappxBidAdapter.js]: Hello @aleksatr sorry about that, we've fixed in [27ad52a7](https://github.com/prebidtappx/Prebid.js/commit/27ad52a706a01503f7406d3661039f06332acd50)

**Comment by aleksatr** [comment]
LGTM

**Pull request merged**